### DISCUSSION TITLE
Installation of plugins docker

### DISCUSSION THREAD

**Discussion by martin323232** [inventree/InvenTree#3954] [Q&A] [answered]
Trying to install Brother plugin.
inventree-brother-plugin is inside plugins.txt file

In terminal: docker-compose run inventree-server invoke install

![Screenshot from 2022-11-18 16-31-34](https://user-images.githubusercontent.com/6439801/202743173-1c782d69-f74b-41a7-a300-90f75c5f17c7.png)
![Screenshot from 2022-11-18 16-32-12](https://user-images.githubusercontent.com/6439801/202743182-a85b0fc8-1bf1-4106-8560-e107460a06ad.png)

Tried installing from docker terminal after installing pip.

How do you install from from webpage, "Install Plugin"?

**Comment by matmair** [comment, marked as answer]
@martin323232 there should appear a modal where you can add the package to your instance if you click `install plugin`. You need superuser privileges for that. Also, you might want to activate the plugin functions - they appear to be off from your screenshot.

Plugins get discovered on server start so maybe also try to restart the server (it should be enough to restart the docker-container for the webserver). Note that plugins are off by default when discovered.

**Reply by SchrodingersGat** [reply]
`docker compose restart` should result in the plugin being discovered

**Reply by martin323232** [reply]
Still no brother plugin after restart.

tried using webpage upload with this as source, https://github.com/inventree/inventree-brother-plugin.git
Nothing after restart.
![Screenshot from 2022-11-19 06-56-10](https://user-images.githubusercontent.com/6439801/202836908-ade711ff-b7b9-49c1-849c-aca5a9b6f230.png)

When trying to install I've tried using pip directly into terminal, in server terminal on docker, in plugin.txt and placing files inside plugin folder.
Is there something I might have messed up by doing that?

**Reply by SchrodingersGat** [reply]
When you install via the web interface (i.e. the "Install Plugin" button) can you confirm that the plugin is entered into your `plugins.txt` file (in the external docker directory)?

**Reply by martin323232** [reply]
No, it is not entered into plugins.txt, it remains empty with only "# InvenTree Plugins (uses PIP framework to install)"

**Reply by SchrodingersGat** [reply]
I would think that this suggests that the external volume is not being mounted as you might expect. Can you try running the command as listed here: https://github.com/inventree/inventree-docs/issues/390

Sequence of commands is:

```
docker compose down
docker volume -d inventree-production_inventree_data
docker compose up -d
```

Then try the plugin sequence again. let me know if this works out for you

**Reply by martin323232** [reply]
inven@inven:~/Downloads/InvenTree-master/docker/production$ sudo docker compose down
[sudo] password for inven: 
[+] Running 5/5
 ⠿ Container inventree-worker            Removed                                                                                    0.0s
 ⠿ Container inventree-proxy             Removed                                                                                    0.0s
 ⠿ Container inventree-server            Removed                                                                                    0.0s
 ⠿ Container inventree-db                Removed                                                                                    0.0s
 ⠿ Network inventree-production_default  Removed                                                                                    0.1s
inven@inven:~/Downloads/InvenTree-master/docker/production$ sudo docker volume -d inventree-production_inventree_data
unknown shorthand flag: 'd' in -d
See 'docker volume --help'.

Usage:  docker volume COMMAND

Manage volumes

Commands:
  create      Create a volume
  inspect     Display detailed information on one or more volumes
  ls          List volumes
  prune       Remove all unused local volumes
  rm          Remove one or more volumes

Run 'docker volume COMMAND --help' for more information on a command.

inven@inven:~/Downloads/InvenTree-master/docker/production$

**Reply by SchrodingersGat** [reply]
Try `docker volume rm -f inventree-production_inventree_data` instead?

**Comment by martin323232** [comment]
That gave another error. Should I do a reinstall and try without using docker?
There is something strange with docker desktop, I've set it to use desktop-linux but containers are not shown in Docker desktop application.

docker volume rm -f inventree-production_inventree_data
Error response from daemon: remove inventree-production_inventree_data: volume is in use - [466a2d959ad4f343031ba5480851d42f1b42a1943a1c84997787e2b779839183, d177217019613c90cb34f24e73a25c7a10c9f342cc7094d3a16b4c0b6f4e13e2, 2b128fe0a62506863db14010c4ada3927eaa62e469da6c182170a01876ba9480, a8450941752b2a7ba98cb91fb3ec9e109fdc84ff80812e9400c1c8be06dcd032, d7f3b73c21acdcdd2d1ef0b1f68c29460a612536c78f99bfed60e1f57d6d87f5, 35834cd39cc4585de4afa7a3ce57e04ceb5a59d9f833b0a1a9d0925fc29a4b67, 7b9a47c4f4681aa03346666f53734707a8e35e3410629def47b3af6bf21a4980, adbb8d1b3fdfb009aa40c316fa03ad739d0c3e307f37a458622f1f866f8ebdb3, a3e32f264cb92aca4d97d5044075e067b9447436c72853c33da91a3cd9ee66c0]

**Reply by wolflu05** [reply]
That means you have not shut down the containers correctly or you have an second folder where you tried the exact same docker compose file. Are you sure you did the docker-compose down thing? Can you show us the output of `docker ps -a`? (Please use markdown formatting of a code block for your output. Three ` in one line, then your output and then again three of these back ticks.

**Reply by martin323232** [reply]
```inven@inven:~/Downloads/InvenTree-master/docker/production$ sudo docker compose down
inven@inven:~/Downloads/InvenTree-master/docker/production$ docker ps -a
CONTAINER ID   IMAGE                        COMMAND                  CREATED        STATUS                      PORTS     NAMES
7b9a47c4f468   inventree/inventree:stable   "/bin/bash ./init.sh…"   46 hours ago   Exited (0) 46 hours ago               inventree-production_inventree-server_run_d4d627fefc8b
466a2d959ad4   inventree/inventree:stable   "/bin/bash ./init.sh…"   46 hours ago   Exited (0) 46 hours ago               inventree-production_inventree-server_run_113892ddf51d
d7f3b73c21ac   inventree/inventree:stable   "/bin/bash ./init.sh…"   46 hours ago   Exited (0) 46 hours ago               inventree-production_inventree-server_run_c22a29855f29
35834cd39cc4   inventree/inventree:stable   "/bin/bash ./init.sh…"   46 hours ago   Exited (0) 46 hours ago               inventree-production_inventree-server_run_caf74188b248
adbb8d1b3fdf   inventree/inventree:stable   "/bin/bash ./init.sh…"   46 hours ago   Exited (0) 46 hours ago               inventree-production_inventree-server_run_36f365f421e2
d17721701961   inventree/inventree:stable   "/bin/bash ./init.sh…"   46 hours ago   Exited (0) 46 hours ago               inventree-production_inventree-server_run_3c0bbb870630
a3e32f264cb9   inventree/inventree:stable   "/bin/bash ./init.sh…"   47 hours ago   Exited (0) 47 hours ago               inventree-production_inventree-server_run_a27b41949e8f
a8450941752b   inventree/inventree:stable   "/bin/bash ./init.sh…"   47 hours ago   Exited (0) 47 hours ago               inventree-production_inventree-server_run_6aeef5c86666
2b128fe0a625   inventree/inventree:stable   "/bin/bash ./init.sh…"   47 hours ago   Exited (137) 47 hours ago             inventree-production_inventree-server_run_befd67a65c7a
inven@inven:~/Downloads/InvenTree-master/docker/production$ ```

**Reply by martin323232** [reply]
I see that there is opended an issue regarding this post. 
So for now running Inventree on docker in Linux it's not possible to use plugins?

Plugins look to be working on Docker for Windows. I've set up a test server there.

**Reply by SchrodingersGat** [reply]
> Plugins look to be working on Docker for Windows. I've set up a test server there.

This is unexpected - the docker image should work identically for windows and linux

**Reply by SchrodingersGat** [reply]
@martin323232 ok, I completely neglected to mention this setting which you need to set in production docker image:

![image](https://user-images.githubusercontent.com/10080325/203014859-36b48cce-266c-4b99-b5a1-5d236490c1e6.png)

Please check this box, and restart the docker container. The plugin *should* be installed then. Sorry, I had forgotten about this setting. Will have to improve the documentation.

**Reply by matmair** [reply]
@SchrodingersGat should we move this to on y default with docker?

**Reply by SchrodingersGat** [reply]
@matmair good idea. I'm currently updating the docs, but "default on" in docker mode would be a good move

**Reply by SchrodingersGat** [reply]
I'll submit a PR for this shortly

**Reply by martin323232** [reply]
It's working now, thanks for your help

**Reply by SchrodingersGat** [reply]
@martin323232 thanks for letting us know. Got there in the end :) The documentation now covers this much more clearly!
